File: algorithms/A_star.py
from heapq import heappush, heappop
from typing import List, Optional, Tuple, Any, Callable
import networkx as nx

def astar(
    G: nx.Graph,
    source: Any,
    target: Any,
    heuristic: Optional[Callable[[Any, Any], float]] = None
) -> Optional[Tuple[List[Any], float, List[Any]]]:
    """
    A* Search algorithm implementation.

    Parameters
    ----------
    G : nx.Graph
        Đồ thị có thể là có trọng số. 
    source : Any
        Đỉnh bắt đầu.
    target : Any
        Đỉnh đích.
    heuristic : Callable[[Any, Any], float], optional
        Hàm heuristic h(n): ước lượng chi phí từ n -> target.

    Returns
    -------
    (path, cost, visited_order)
        path : danh sách các node trên đường đi ngắn nhất.
        cost : tổng chi phí đường đi.
        visited_order : thứ tự mở rộng node.
    """
    if heuristic is None:
        heuristic = lambda u, v: 0  # nếu không có heuristic thì trở thành Dijkstra

    open_set = []
    heappush(open_set, (0 + heuristic(source, target), 0, source))  # (f = g+h, g, node)
    came_from = {}
    g_score = {source: 0}
    visited_order = []

    while open_set:
        f, g, current = heappop(open_set)
        visited_order.append(current)

        if current == target:
            # reconstruct path
            path = [current]
            while current in came_from:
                current = came_from[current]
                path.append(current)
            path.reverse()
            return path, g_score[target], visited_order

        for neighbor in G.neighbors(current):
            weight = G[current][neighbor].get('weight', 1)
            tentative_g = g_score[current] + weight
            if tentative_g < g_score.get(neighbor, float('inf')):
                came_from[neighbor] = current
                g_score[neighbor] = tentative_g
                f_score = tentative_g + heuristic(neighbor, target)
                heappush(open_set, (f_score, tentative_g, neighbor))

    return None


if __name__ == "__main__":
    G = nx.Graph()
    G.add_node('A', pos=(0, 0))
    G.add_node('B', pos=(1, 2))
    G.add_node('C', pos=(2, 0))
    G.add_node('D', pos=(4, 1))

    # Các cạnh có trọng số
    edges = [
        ('A', 'B', 2.5),
        ('A', 'C', 2.0),
        ('B', 'D', 3.0),
        ('C', 'D', 2.2)
    ]
    G.add_weighted_edges_from(edges)
    
    def heuristic(u, v):
        pos_u = G.nodes[u].get('pos')
        pos_v = G.nodes[v].get('pos')
        if pos_u is None or pos_v is None:
            return 0.0
        (x1, y1), (x2, y2) = pos_u, pos_v
        return ((x1 - x2)**2 + (y1 - y2)**2)**.5

    # Chạy A*
    result = astar(G, 'A', 'D', heuristic)
    print("Path:", result[0])
    print("Cost:", result[1])
    print("Visited:", result[2])

File: webapp/src/algorithms/astar.js
// A* Search algorithm with step-by-step visualization

import { getNeighbors, euclideanDistance } from '../utils/graphUtils';

/**
 * Min-heap priority queue implementation
 */
class PriorityQueue {
  constructor() {
    this.heap = [];
  }

  push(item, priority) {
    this.heap.push({ item, priority });
    this.bubbleUp(this.heap.length - 1);
  }

  pop() {
    if (this.heap.length === 0) return null;
    const top = this.heap[0];
    const last = this.heap.pop();
    if (this.heap.length > 0) {
      this.heap[0] = last;
      this.bubbleDown(0);
    }
    return top;
  }

  isEmpty() {
    return this.heap.length === 0;
  }

  bubbleUp(index) {
    while (index > 0) {
      const parent = Math.floor((index - 1) / 2);
      if (this.heap[parent].priority <= this.heap[index].priority) break;
      [this.heap[parent], this.heap[index]] = [this.heap[index], this.heap[parent]];
      index = parent;
    }
  }

  bubbleDown(index) {
    while (true) {
      const left = 2 * index + 1;
      const right = 2 * index + 2;
      let smallest = index;
      if (left < this.heap.length && this.heap[left].priority < this.heap[smallest].priority) {
        smallest = left;
      }
      if (right < this.heap.length && this.heap[right].priority < this.heap[smallest].priority) {
        smallest = right;
      }
      if (smallest === index) break;
      [this.heap[smallest], this.heap[index]] = [this.heap[index], this.heap[smallest]];
      index = smallest;
    }
  }

  toArray() {
    return this.heap.map(h => h.item);
  }
}

/**
 * A* Search algorithm implementation with event emission for visualization
 * @param {Object} graph - Graph object with nodes and edges
 * @param {string} source - Source node id
 * @param {string} target - Target node id
 * @param {boolean} isDirected - Whether the graph is directed
 * @param {Object} options - Options object (can contain custom heuristic)
 * @returns {Object} - Result with path, cost, steps, and visited order
 */
export function astar(graph, source, target, isDirected = false, options = {}) {
  const steps = [];
  const gScore = new Map(); // Cost from start
  const fScore = new Map(); // Estimated total cost
  const parent = new Map();
  const visited = new Set();
  const pq = new PriorityQueue();
  const visitOrder = [];

  // Extract heuristic from options or use default
  let heuristic = typeof options === 'function' ? options : options?.heuristic;
  
  // Default heuristic: Euclidean distance
  if (!heuristic || typeof heuristic !== 'function') {
    heuristic = (node) => euclideanDistance(graph, node, target) * 0.5;
  }

  // Initialize
  for (const [id] of graph.nodes) {
    gScore.set(id, Infinity);
    fScore.set(id, Infinity);
  }

  gScore.set(source, 0);
  fScore.set(source, heuristic(source));
  parent.set(source, null);
  pq.push(source, fScore.get(source));

  // Initial step
  steps.push({
    type: 'init',
    current: null,
    gScores: Object.fromEntries(gScore),
    fScores: Object.fromEntries(fScore),
    visited: [...visited],
    frontier: pq.toArray(),
    parent: Object.fromEntries(parent),
    message: `Starting A* from node ${source} to ${target}`,
  });

  while (!pq.isEmpty()) {
    const { item: current } = pq.pop();

    if (visited.has(current)) continue;

    visited.add(current);
    visitOrder.push(current);

    const g = gScore.get(current);
    const f = fScore.get(current);
    const h = f - g;

    // Expand step
    steps.push({
      type: 'expand',
      current,
      gScore: g,
      hScore: h,
      fScore: f,
      gScores: Object.fromEntries(gScore),
      fScores: Object.fromEntries(fScore),
      visited: [...visited],
      frontier: pq.toArray(),
      parent: Object.fromEntries(parent),
      message: `Expanding ${current}: g=${g.toFixed(1)}, h=${h.toFixed(1)}, f=${f.toFixed(1)}`,
    });

    // Check if we reached the target
    if (current === target) {
      const path = reconstructPath(parent, target);
      const cost = gScore.get(target);

      steps.push({
        type: 'goal',
        current,
        path,
        cost,
        gScores: Object.fromEntries(gScore),
        visited: [...visited],
        message: `Found target ${target}! Path cost: ${cost.toFixed(1)}`,
      });

      return {
        found: true,
        path,
        cost,
        steps,
        visitOrder,
        nodesVisited: visited.size,
      };
    }

    // Explore neighbors
    const neighbors = getNeighbors(graph, current, isDirected);
    for (const { node: neighbor, weight } of neighbors) {
      if (visited.has(neighbor)) continue;

      const tentativeG = gScore.get(current) + weight;

      if (tentativeG < gScore.get(neighbor)) {
        const oldG = gScore.get(neighbor);
        const h = heuristic(neighbor);

        parent.set(neighbor, current);
        gScore.set(neighbor, tentativeG);
        fScore.set(neighbor, tentativeG + h);
        pq.push(neighbor, tentativeG + h);

        steps.push({
          type: 'relax',
          current,
          neighbor,
          oldGScore: oldG,
          newGScore: tentativeG,
          hScore: h,
          fScore: tentativeG + h,
          weight,
          gScores: Object.fromEntries(gScore),
          fScores: Object.fromEntries(fScore),
          visited: [...visited],
          frontier: pq.toArray(),
          parent: Object.fromEntries(parent),
          message: `Relaxing ${current}→${neighbor}: g=${tentativeG.toFixed(1)}, h=${h.toFixed(1)}, f=${(tentativeG + h).toFixed(1)}`,
        });
      }
    }
  }

  // No path found
  steps.push({
    type: 'no_path',
    visited: [...visited],
    message: `No path found from ${source} to ${target}`,
  });

  return {
    found: false,
    path: [],
    cost: Infinity,
    steps,
    visitOrder,
    nodesVisited: visited.size,
  };
}

/**
 * Reconstruct path from parent map
 */
function reconstructPath(parent, target) {
  const path = [];
  let current = target;
  while (current !== null) {
    path.unshift(current);
    current = parent.get(current);
  }
  return path;
}

export default astar;
